File: backend/src/routes/animais.ts
import { Hono } from "hono";
import { readDB, writeDB, nextId } from "../db/json.ts";
import { getMetricsAnimal, getDashboardKPIs } from "../services/metrics.ts";
import { calcularGMD } from "../services/calculos.ts";
import { formatDate } from "../services/calculos.ts";

const animais = new Hono();

// Listar animais
animais.get("/", async (c) => {
  const db = await readDB();
  const { status, lote_id } = c.req.query();
  
  let filtered = db.animais;
  
  if (status) {
    filtered = filtered.filter((a: any) => a.status === status);
  }
  
  if (lote_id) {
    filtered = filtered.filter((a: any) => a.lote_id === parseInt(lote_id));
  }
  
  // Adicionar métricas
  const result = [];
  for (const animal of filtered) {
    const metrics = await getMetricsAnimal(animal.id);
    const lote = db.lotes.find((l: any) => l.id === animal.lote_id);
    
    result.push({
      ...animal,
      peso_atual: metrics?.peso_atual || animal.peso_entrada,
      gmd: metrics?.gmd > 0 ? metrics.gmd : null,
      gmd_status: metrics?.gmd_status || null,
      lote_nome: lote?.nome || null
    });
  }
  
  return c.json(result);
});

// Buscar animal por ID
animais.get("/:id", async (c) => {
  const id = parseInt(c.req.param("id"));
  const db = await readDB();
  const animal = db.animais.find((a: any) => a.id === id);
  
  if (!animal) {
    return c.json({ error: "Animal não encontrado" }, 404);
  }
  
  const metrics = await getMetricsAnimal(animal.id);
  
  return c.json({
    ...animal,
    metrics
  });
});

// Criar animal
animais.post("/", async (c) => {
  const body = await c.req.json();
  
  // Validações obrigatórias
  if (!body.brinco) {
    return c.json({ error: "Brinco é obrigatório" }, 400);
  }
  if (!body.data_entrada) {
    return c.json({ error: "Data de entrada é obrigatória" }, 400);
  }
  if (!body.peso_entrada || body.peso_entrada <= 0) {
    return c.json({ error: "Peso de entrada é obrigatório" }, 400);
  }
  
  // Validação: data não pode ser futura
  const dataEntrada = new Date(body.data_entrada);
  const hoje = new Date();
  hoje.setHours(23, 59, 59, 999);
  if (dataEntrada > hoje) {
    return c.json({ error: "Data de entrada não pode ser futura" }, 400);
  }
  
  // Validação: peso não pode ser negativo
  if (body.peso_entrada < 0) {
    return c.json({ error: "Peso não pode ser negativo" }, 400);
  }
  
  // Validação: composição cruzado não pode passar de 100%
  if (body.raca === 'Cruzado' && body.composicao) {
    const total = body.composicao.reduce((sum: number, c: any) => sum + (c.porcentagem || 0), 0);
    if (total > 100) {
      return c.json({ error: `Porcentagem total (${total}%) não pode passar de 100%` }, 400);
    }
  }
  
  const db = await readDB();
  
  // Verifica brinco duplicado
  if (db.animais.find((a: any) => a.brinco === body.brinco)) {
    return c.json({ error: "Brinco já cadastrado" }, 400);
  }
  
  const id = await nextId("animalId");
  
  const novoAnimal = {
    id,
    brinco: body.brinco,
    raca: body.raca || "Nelore",
    composicao: body.composicao || null,
    sexo: body.sexo || "MACHO",
    data_entrada: body.data_entrada,
    peso_entrada: body.peso_entrada,
    lote_id: body.lote_id || null,
    observacao: body.observacao || null,
    status: body.status || "ATIVO",
    created_at: new Date().toISOString(),
    updated_at: new Date().toISOString()
  };
  
  db.animais.push(novoAnimal);
  
  // Cria pesagem inicial automática
  db.pesagens.push({
    id: await nextId("pesagemId"),
    animal_id: id,
    data_pesagem: body.data_entrada,
    peso: body.peso_entrada,
    tecnico: "Sistema",
    created_at: new Date().toISOString()
  });
  
  await writeDB(db);
  
  const metrics = await getMetricsAnimal(id);
  
  return c.json({
    ...novoAnimal,
    metrics
  }, 201);
});

// Atualizar animal
animais.put("/:id", async (c) => {
  const id = parseInt(c.req.param("id"));
  const body = await c.req.json();
  const db = await readDB();
  
  const index = db.animais.findIndex((a: any) => a.id === id);
  if (index === -1) {
    return c.json({ error: "Animal não encontrado" }, 404);
  }
  
  // Verifica brinco duplicado se mudou
  if (body.brinco && body.brinco !== db.animais[index].brinco) {
    if (db.animais.find((a: any) => a.brinco === body.brinco && a.id !== id)) {
      return c.json({ error: "Brinco já cadastrado" }, 400);
    }
  }
  
  db.animais[index] = {
    ...db.animais[index],
    ...body,
    updated_at: new Date().toISOString()
  };
  
  await writeDB(db);
  
  const metrics = await getMetricsAnimal(id);
  
  return c.json({
    ...db.animais[index],
    metrics
  });
});

// Deletar animal
animais.delete("/:id", async (c) => {
  const id = parseInt(c.req.param("id"));
  const db = await readDB();
  
  const index = db.animais.findIndex((a: any) => a.id === id);
  if (index === -1) {
    return c.json({ error: "Animal não encontrado" }, 404);
  }
  
  // Remove pesagens
  db.pesagens = db.pesagens.filter((p: any) => p.animal_id !== id);
  
  // Remove animal
  db.animais.splice(index, 1);
  
  await writeDB(db);
  
  return c.body(null, 204);
});

export default animais;

File: backend/app/api/endpoints/animais.py
"""Endpoints de Animais."""
from typing import List, Optional
from datetime import date
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session

from app.core import get_db
from app.models import Animal, Lote, Pesagem
from app.schemas import AnimalCreate, AnimalUpdate, AnimalResponse, AnimalListResponse, MetricasAnimal
from app.services import AnimalService

router = APIRouter(prefix="/animais", tags=["Animais"])


@router.get("", response_model=List[AnimalListResponse])
def listar_animais(
    status: Optional[str] = Query(None),
    lote_id: Optional[int] = Query(None),
    db: Session = Depends(get_db)
):
    """Lista todos os animais."""
    query = db.query(Animal).join(Lote, Animal.lote_id == Lote.id, isouter=True)

    if status:
        query = query.filter(Animal.status == status)
    if lote_id:
        query = query.filter(Animal.lote_id == lote_id)

    animais = query.order_by(Animal.brinco).all()

    result = []
    for animal in animais:
        peso_atual = AnimalService.get_peso_atual(db, animal.id)
        metrics = AnimalService.get_metricas(db, animal)

        result.append(AnimalListResponse(
            id=animal.id,
            brinco=animal.brinco,
            raca=animal.raca,
            sexo=animal.sexo,
            status=animal.status,
            peso_entrada=animal.peso_entrada,
            data_entrada=animal.data_entrada,
            lote_id=animal.lote_id,
            lote_nome=animal.lote.nome if animal.lote else None,
            peso_atual=peso_atual,
            gmd=metrics["gmd"] if metrics["gmd"] > 0 else None,
            gmd_status=metrics["gmd_status"] if metrics["gmd"] > 0 else None
        ))

    return result


@router.get("/{animal_id}", response_model=AnimalResponse)
def buscar_animal(animal_id: int, db: Session = Depends(get_db)):
    """Busca um animal específico com métricas."""
    animal = db.query(Animal).filter(Animal.id == animal_id).first()
    if not animal:
        raise HTTPException(status_code=404, detail="Animal não encontrado")

    metrics = AnimalService.get_metricas(db, animal)
    animal_dict = {
        "id": animal.id,
        "brinco": animal.brinco,
        "raca": animal.raca,
        "sexo": animal.sexo,
        "data_entrada": animal.data_entrada,
        "peso_entrada": animal.peso_entrada,
        "lote_id": animal.lote_id,
        "observacao": animal.observacao,
        "status": animal.status,
        "created_at": animal.created_at,
        "updated_at": animal.updated_at,
        "metrics": MetricasAnimal(**metrics)
    }
    return animal_dict


@router.post("", response_model=AnimalResponse, status_code=status.HTTP_201_CREATED)
def criar_animal(animal: AnimalCreate, db: Session = Depends(get_db)):
    """Cria um novo animal."""
    # Verifica se o brinco já existe
    existente = db.query(Animal).filter(Animal.brinco == animal.brinco).first()
    if existente:
        raise HTTPException(status_code=400, detail="Brinco já cadastrado")

    # Verifica se lote existe (se fornecido)
    if animal.lote_id:
        lote = db.query(Lote).filter(Lote.id == animal.lote_id).first()
        if not lote:
            raise HTTPException(status_code=400, detail="Lote não encontrado")

    db_animal = Animal(**animal.model_dump())
    db.add(db_animal)
    db.commit()
    db.refresh(db_animal)

    # Cria pesagem inicial automaticamente
    primeira_pesagem = Pesagem(
        animal_id=db_animal.id,
        data_pesagem=animal.data_entrada,
        peso=animal.peso_entrada,
        tecnico="Sistema"
    )
    db.add(primeira_pesagem)
    db.commit()

    metrics = AnimalService.get_metricas(db, db_animal)
    db_animal_dict = {
        "id": db_animal.id,
        "brinco": db_animal.brinco,
        "raca": db_animal.raca,
        "sexo": db_animal.sexo,
        "data_entrada": db_animal.data_entrada,
        "peso_entrada": db_animal.peso_entrada,
        "lote_id": db_animal.lote_id,
        "observacao": db_animal.observacao,
        "status": db_animal.status,
        "created_at": db_animal.created_at,
        "updated_at": db_animal.updated_at,
        "metrics": MetricasAnimal(**metrics)
    }
    return db_animal_dict


@router.put("/{animal_id}", response_model=AnimalResponse)
def atualizar_animal(animal_id: int, data: AnimalUpdate, db: Session = Depends(get_db)):
    """Atualiza um animal."""
    animal = db.query(Animal).filter(Animal.id == animal_id).first()
    if not animal:
        raise HTTPException(status_code=404, detail="Animal não encontrado")

    update_data = data.model_dump(exclude_unset=True)
    
    # Se mudou o brinco, verifica se não existe
    if "brinco" in update_data:
        existente = db.query(Animal).filter(
            Animal.brinco == update_data["brinco"],
            Animal.id != animal_id
        ).first()
        if existente:
            raise HTTPException(status_code=400, detail="Brinco já cadastrado")

    for key, value in update_data.items():
        setattr(animal, key, value)

    db.commit()
    db.refresh(animal)

    metrics = AnimalService.get_metricas(db, animal)
    animal_dict = {
        "id": animal.id,
        "brinco": animal.brinco,
        "raca": animal.raca,
        "sexo": animal.sexo,
        "data_entrada": animal.data_entrada,
        "peso_entrada": animal.peso_entrada,
        "lote_id": animal.lote_id,
        "observacao": animal.observacao,
        "status": animal.status,
        "created_at": animal.created_at,
        "updated_at": animal.updated_at,
        "metrics": MetricasAnimal(**metrics)
    }
    return animal_dict


@router.delete("/{animal_id}", status_code=status.HTTP_204_NO_CONTENT)
def deletar_animal(animal_id: int, db: Session = Depends(get_db)):
    """Deleta um animal e suas pesagens."""
    animal = db.query(Animal).filter(Animal.id == animal_id).first()
    if not animal:
        raise HTTPException(status_code=404, detail="Animal não encontrado")

    # Deleta pesagens primeiro
    db.query(Pesagem).filter(Pesagem.animal_id == animal_id).delete()
    db.delete(animal)
    db.commit()
    return None